man/email2ncid.conf.5
.\" %W% %G%
.TH EMAIL2NCID.CONF 5
.SH NAME
.B email2ncid.conf
- email2ncid configuration file
.SH DESCRIPTION
The \fIemail2ncid.conf\fR file contains the configuration information for
\fIemail2ncid\fR.
.PP
The \fIemail2ncid.conf\fR file understands 3 types of lines:
.TP
.B blank line
ignored
.TP
.B comment line, beginning with '#'
ignored
.TP
.B VARIABLE=VALUE
VALUE can be between quotes.
.SH OPTIONS
.TP
.B NCIDSERVER=<host address>:<port>
The NCID server host address can be an IP address or a resolvable host name.

Default: ncidserver = localhost:3333
.TP
.B lineid
Suggested values: WIRELESS NETWORK Internet

Default: lineid = NETWORK
.PP
See email2ncid.conf for a complete list of variables to set and
what the default values are.
.SH EXAMPLES
Set the NCID server IP address: ncidserver = 192.168.1.20
.PP
Set the NCID server port: ncidserver = :3333
.SH SEE ALSO
email2ncid.1,
procmail.1


doc/CHANGES
ncid - Network CallerID

Version 1.6
    Makefile: removed comment: #-Wunused-but-set-variable -Wmaybe-uninitialized
    - client/ncid.sh: changed: typeWidth to mtypeWidth
                      added: labelWidth dateWidth timeWidth
                      changed: historyTextWidth from 86 to an expression
                      added: doVerbose Mesg Type Width
                      added set lbl5 and lbl8
                      fieldInfo {} new
                      serverOPT {} modified text
                      connectCID {} added: 'HELLO: IDENT' and 'HELLO: CMD'
                          changed: historyTextWidth from 86 to sum of fields
                          added: labelWidth dateWidth timeWidth spacesWidth
                          renamed: typeWidth to mtypeWidth
                      makeWindow {} changed history front from 11 to 12
                            made history window width dependent on time format
                            added $lbl5, $lbl8
                            added label for fields as row 1
                            added increased other grid rows by 1
                            added "set nmbrREQ [string trimleft $nmbrREQ]"
                            added "Field Labekls" menu item
                      logClock {} added variable historyTextWidth
                            add: + 3 to historywidth if 12 hour time display
                            add: - 3 to historywidth if 24 hour time display
                            add: configure window width, set geometry
                            add: change minimum window size, update
                            added: $lbl5 and $lbl8 lines
                      displayLOG {} moved display format here from ????Width
    - doc/Verbose.md: updated
          NCID-API.md: fixed color problem; add HELLO: lines
          Message.md: add HELLO: lines
    - gateway/email2ncid.pl: fixed default port from 3334 to 3333
                            added: $IDENT and $COMMAND
              ncid2ncid.c: socketConnect():
                             added: IDENT and COMMAND
                             changed if statement to find 2x0, and 25x lines
              ncid2ncid.h: added: #define IDENT and #define COMMAND
              obi2ncid.pl: added: $IDENT and $COMMAND
                                  if (/^[23]\d\d/) { logMsg(1, $_); }
                                  else { logMsg(7, $_); }
              rn2ncid.pl: added: $IDENT and $COMMAND
                                 if (/^[23]\d\d/) { logMsg(1, $_); }
                                 else { logMsg(5, $_); }
              sip2ncid.c: socketConnect()
                            added IDENT and COMMAND
                            added code to read lines up to end of startup
                          readLine(): new
              sip2ncid.h: added: #define IDENT and #define COMMAND
              wc2ncid.pl: added: $IDENT and $COMMAND
                                 if (/^[23]\d\d/) { logMsg(1, $_); }
                                 else { logMsg(5, $_); }
              yac2ncid.sh: added: $IDENT and $COMMAND
    - man/ncidd.conf.5: added HELLO
              added several missing settings e.g., announce, audiofmt, and more
    - server/Makefile: added -Wmisleading-indentation
    - server/ncidd.c: added doHELLO() and notHex()
               added ident to struct ipinfo
               doPoll() fixed if test:
                 added IPinfo[pos].ident to some connect/disconnect messages
                 changed: if ((buf[0] == '+' && buf[4] == ':') || buf[8] == ':')
                 to: if (buf[0] == '+' && (buf[4] == ':' || buf[8] == ':'))
                 changed: else if (cid.status == (CIDALL3 | CIDMESG) )
                 to: else if (notHex() && cid.status == (CIDALL3 | CIDMESG))
               showConnected() added IPinfo[pos].ident to log messages
             nciddalias.c: strmatch() changed inside last else
                           from: len = strlen(string);
                           to:   len = strlen(value);  
    - screenshots/ncid-call.png: updated
    - screenshots/ncid-mesg.png: updated
    - screenshots/ncid-FieldLabels.png: added
    - screenshots/ncid-ServerMenuHelp.png: added
    - screenshots/ncid-ServerOptions.png: added
    - server/ncidd.h: added #define HELLO, #define IDENT, & #define COMMAND
    - server/nciddhangup.c: hangupCall() added missing arentheses to else
                                         inside if (!ttyfd) {...}
    - test/test-server: added 407 alias
           test-gw-calls.data: added CALL and CALLINFO lines for number 407
           renamed test-script.data to test-mobile.data
           test-client-screenshot.data: updated
           test-client.data: updated

Version 1.5
    - Makefile: added doc/NCID-SDK.tar.gz to DOC
                added setup/ to subdirectories
                added logrotatedir to freebsd
                added install-logrotate to install-freebsd
                removed MFLAGS and LDFLAGS from cygwin target
    - client/ncid.sh: getCID{} added "set argument ..." in if {$mod_menu}
                      removed: package require tile
                      changed maxNumber Width to numberWidth
                      changed numberWidth from 14 to 20
                      added nameWidth = 20 was hardcoded at 18
                      added typeWidth = 7
                      changed historyTextWidth from 81 to 86 
                      display all *width values at verbose 1
                      allow selection of MSG and NOT lines
    - debian/Makefile: updated
             README.debian: updated
             ncid-mysql.init: new
    - doc/Makefile: updated and removed target sdk from the default
                    changed "cp -a" to "cp -Rp"
          Alias.md: added information about a name/number like "#########8"
          mk-epub: modified code for getting options
          mk-manual-date: added ':' case statement in getopts()
                          modified options code

gateway/email2ncid.pl
#!/usr/bin/perl

# email2ncid - email to NCID gateway

# Requires: procmail

# The email2ncid script is called by procmail when an NCID email
# is received.  It is not a daemon like the other gateways.

# If an email subject line is "NCID Message", the email is converted
# into an NCID message (MSG:) and sent to the server.

# Copyright (c) 2016
#  by John L. Chmielewski <jlc@users.sourceforge.net>

# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

use strict;
use warnings;
use Getopt::Long qw(:config no_ignore_case_always);
use File::Basename;
use Config::Simple;
use Pod::Usage;
use IO::Socket::INET;

my $prog = basename($0);
my $confile = basename($0, '.pl');
my $VERSION = "(NCID) XxXxX";

my $IDENT = "HELLO: IDENT: gateway $prog $VERSION";
my $COMMAND = "HELLO: CMD: no_log";

my $ConfigDir = "/usr/local/etc/ncid";
my $ConfigFile = "$ConfigDir/$confile.conf";

my ($help, $man, $version);

my ($ncidaddr, $ncidport) = ("localhost", 3333);
my $NCIDserver = undef;
my $lineid = "NETWORK";
my $lineID = undef;
my %config;
my ($ncidsock, $ncidserver, $tmp, $test, $cfg);
my ($subj, $start, $plain, $html, $multi, $meta, $status, $notify);
my ($from, $addr, $boundary, $message);

$test = $subj = $start = $plain = $html = $multi = $meta = $status = 0;
$notify = 0;
$ncidserver = $boundary = $message = $from = $addr = "";

Getopt::Long::Configure ("bundling");
my ($result) = GetOptions("ncidserver|n=s" => \$NCIDserver,
               "configfile|C=s" => \$ConfigFile,
               "help|h" => \$help,
               "notify|N" => \$notify,
               "man|m" => \$man,
               "version|V" => \$version,
               "test|t=i" => \$test,
             ) || pod2usage(2);
die "$prog $VERSION\n" if $version;
pod2usage(-verbose => 1, -exitval => 0) if $help;
pod2usage(-verbose => 2, -exitval => 0) if $man;

# reading configuration file after command line processing
# is necessary because the command line can change the
# location of the configuration file
$cfg = new Config::Simple($ConfigFile);
if (defined $cfg) {
# opened config file
    %config = $cfg->vars();
    $ncidserver = $config{'default.ncidserver'};
    $lineID = $config{'default.lineid'};
}

# these must always have a value
$lineid = $lineID if $lineID;

# these command line values override the configuration file values
$ncidserver = $NCIDserver if $NCIDserver;

if ($ncidserver) {
    ($tmp) = $ncidserver =~ /([^:]*)/;
    $ncidaddr = $tmp if $tmp;
    ($tmp) = $ncidserver =~ /.*:(.*)/;
    $ncidport = $tmp if $tmp;
}

while (<STDIN>) {
    if ($test >= 3) { printf "> $_"; }
    if (/^Subject: /) {
        if (/^Subject: +NCID +Message/) {
            $subj = 1;
        } elsif ($notify) {
            chop;
            ($message) = /^Subject:(.*)/;
            last if $from;
        }
        next;
    }
    if (/^From: /) {
        ($from) = /^From: "?([^"]*?) </;
        ($addr) = / <([^>]*)/;
        next;
    }
    if (/^Content-Type: +text\/plain/) {
        $plain = 1;
        next;
    }
    if (/^Content-Type: +text\/html/) {
        $html = 1;
        next;
    }
    if (/^Content-Type: +multipart\/alternative/) {
        $multi = 1;
    }
    if (/<meta http-equiv=/) {
        $meta = 1;
        next;
    }
    if (/^Status: RO/) {
        $status = 1;
        next;
    }
    if ($subj == 1) {
        # get the boundary characters
        if (/boundary/)
        {
            ($boundary) = /boundary="?([^"\n]*)/;
            next;
        }
        if (/^\s*$/) {
            if ($html== 1 || $multi == 1 || $meta == 1) { $status = 0; }
            # if content plain and status were not found, skip line
            if ($plain == 0 && $status == 0) { next; }
            # first blank line indicates start of text
            if ($start == 0) { $start = 1; }
            next;
        }
        if ($start == 1) {
            chop;
            # boundary may be at end of message
            if ($boundary && $_ =~ $boundary) { last; }
            # combine text lines into one message line
            $message = "$message $_";
            next;
        }
    }
}
if ($test >= 1) {
    print "test=$test\n";
    print "Configuration File: $ConfigFile\n";
    print("subj=$subj start=$start plain=$plain html=$html multi=$multi meta=$meta status=$status\n");
    print "ncidserver: $ncidaddr:$ncidport\n";
}

if (!$test || $test >= 2) { &connectNCID; }
if ($message) {
    $message = "MSG:$message ###NAME*$from*NMBR*$addr*LINE*$lineid*MTYPE*EMAIL";
    if ($test >= 1) {
        print("from=$from  addr=$addr boundary=$boundary\n");
        print "$message\n";
    }
    else { print($ncidsock "$message\r\n"); }
}
close($ncidsock) if $ncidsock;

sub connectNCID {
  my $api = "";
  my $ignore;

  $ncidsock = IO::Socket::INET->new(
    Proto    => "tcp",
    PeerAddr => $ncidaddr,
    PeerPort => $ncidport,
  );

  # $ncidsock undefined if could not connect to server
  if (!$ncidsock) {
    print("NCID server: $ncidaddr:$ncidport $!");
    exit 1;
  }

  # send ident to server
  print $ncidsock "$IDENT\n";

  # make sure call log not sent
  print $ncidsock "$COMMAND\n";

  if ($test >= 1) {
    print "    Sent: $IDENT\n";
    print "    Sent: $COMMAND\n";
  }

  my $greeting = <$ncidsock>;

  # read and discard cidcall log sent from server
  while (<$ncidsock>)
  {
    # a log file may or may not be sent
    # but a 300 message is always sent
    $api = $_ if /^210/;
    last if /^300/;
    $ignore = $_;
  };
  if ($test >= 2) {
    print("    $greeting");
    print("    $api") if $api;
    print("    $ignore"); # 300 message
  }
}

=head1 NAME

email2ncid - convert an email to an NCID message

=head1 SYNOPSIS

 email2ncid [--configfile|-C <filename>]
            [--help|-h]
            [--man|-m]
            [--notify|-N]
            [--ncidserver|-n <[host][:port]>]
            [--test|-t <1-9>]
            [--version|-V]

=head1 DESCRIPTION

The email2ncid gateway sends the contents of an email to the NCID
server as one line.  It is called from a .procmailrc file when
an email contains the line: Subject: NCID Message

The email2ncid gateway has an option to only send an email subject
line to the NCID server.  It is called from a .procmailrc file when
the email address or name matches on the email "From:" line.

The email must be in either plain text, or HTML and plain text.
The output of email2ncid is a one line NCID message sent to
an NCID server.

=head1 OPTIONS

=over

=item -C, --configfile <filename>

Specifies the configuration file to use.  The program will still run if
a configuration file is not found.

=item -h, --help

Displays the help message and exits.

=item -m, --man

Displays the manual page and exits.

=item -N, --notify

This option sends a message to NCID containing only the subject line
instead of the email contents as one line.  It is a notification of
some type:

    * visitor arrived at a gate in a gated community
    * an important email arrived

=item -n <[host][:port]>, --ncidserver=<[host][:port]>

Specifies the NCID server.
Port may be specified by suffixing the hostname with :<port>.

Input must be <host:port> or <host>, or <:host>

=item -t, --test <1-9>

Test mode connects to the server and displays some information
and the message.  It does not send the message to the server.
Set the level to a higher number for more information.
Levels range from 1 to 9, but not all levels are used.

    test = 1: show some variables and generated message
    test = 2: additionally show 2 or 3 lines returned by server
    test = 3: additionally show tne email message

Default: no test mode

=item -V, --version

Displays the version.

=back

=head1 REQUIREMENTS

=over

=item The NCID server

http://ncid.sourceforge.net/ncid/ncid.html

=item A dynamic DNS service:

ChangeIP (https://www.changeip.com/dns.php)

Dynu     (https://www.dynu.com)

DynDNS   (https://www.dyn.com)

=item A Mail Transport Agent (MTA):

exim, postfix, sendmail, etc.

=item firewall:

Forward port 25 TCP/UDP to the computer running the MTA

=item procmail:

$HOME/.procmailrc must be created or updated to call email2ncid.

Execute the following command to automate this process:

=over

ncid-setup procmailrc

=back

=item Perl

perl 5.6 or higher,
perl(Config::Simple)

=back

=head1 FILES

/etc/ncid/email2ncid.conf

$HOME/.procmailrc

=head1 SEE ALSO

ncidd.8,
email2ncid.conf.5,
procmail.1
